src/main/java/com/github/manolo8/darkbot/gui/drawables/ZonesDrawer.java

```diff
@@ -120,16 +124,31 @@ public void drawCustomZonePath(MapGraphics mg, ZoneInfo zoneInfo) {
     public void drawSafeZone(MapGraphics mg, SafetyInfo safetyInfo) {
         if (safetyInfo == null) return;
 
-        mg.drawOvalCentered(safetyInfo, mg.toScreenPointX(safetyInfo.getDiameter()),
-                mg.toScreenPointY(safetyInfo.getDiameter()), true);
+        mg.drawOvalCentered(safetyInfo, safetyInfo.getDiameter() / mg.getScaleX(),
+                safetyInfo.getDiameter() / mg.getScaleY(), true);
+    }
+
+    private int gridToMapX(MapGraphics mg, int x, int width) {
+        return (int) (mg.toScreenPointX(0) + x * width / zoneResolution.getValue());
     }
 
-    private int gridToMapX(MapGraphics mg, int x) {
-        return x * mg.getWidth() / zoneResolution.getValue();
+    private int gridToMapY(MapGraphics mg, int y, int height) {
+        return (int) (mg.toScreenPointY(0) + y * height / zoneResolution.getValue());
     }
 
-    private int gridToMapY(MapGraphics mg, int y) {
-        return y * mg.getHeight() / zoneResolution.getValue();
+    private int getWidth(MapGraphics mg, double gameWidth) {
+        double minX = mg.toScreenPointX(0);
+        double maxX = mg.toScreenPointX(gameWidth);
+
+        return (int) (maxX - minX);
+    }
+
+    //todo or add those two methods in MapGraphics?
+    private int getHeight(MapGraphics mg, double gameHeight) {
+        double minY = mg.toScreenPointY(0);
+        double maxY = mg.toScreenPointY(gameHeight);
+
+        return (int) (maxY - minY);
     }
```

Comment:
These are basically mg.toScreenSizeW & H i mentioned earlier, they should be going away